R-packages/modeltools/R/matrix.R

```diff
@@ -87,84 +82,133 @@ create_train_and_predict_matrices <- function(lagged_df, ahead, training_window_
                     msg = paste("multiple responses at ahead =", a))
     }
     
+    # prediction matrices are the same for all aheads
+    predict_time_value <- max(lagged_df$time_value)
+    predict_x <- lagged_df %>%
+        filter(time_value == predict_time_value) %>%
+        select(tidyselect::starts_with("value")) %>%
+        as.matrix()
+    predict_geo_values <- lagged_df %>%
+        filter(time_value == predict_time_value) %>%
+        select(geo_value) %>% pull()
+    
+    all_out <- list()
+    
     if (aheads_separate) {
-        for (a in ahead) {
-            out <- list()  # matrices corresponding to ahead+a
-            
-            # Find the last possible date of training data
-            response_end_date <- lagged_df %>%
-                select(time_value, tidyselect::starts_with(sprintf("response+%i:", a))) %>%
-                tidyr::drop_na() %>%
-                summarize(max(time_value)) %>%
-                pull()
-            train_end_date <- min(max(lagged_df$time_value), response_end_date)
-            
-            # Training matrices
-            out$train_x <- lagged_df %>%
-                select(geo_value, time_value, tidyselect::starts_with("value")) %>%
-                filter(between(time_value,
-                               train_end_date - training_window_size + 1,
-                               train_end_date)) %>%
-                select(-c(geo_value, time_value)) %>%
-                as.matrix()
-            out$train_y <- lagged_df %>%
-                filter(between(time_value,
-                               train_end_date - training_window_size + 1,
-                               train_end_date)) %>%
-                select(tidyselect::starts_with(sprintf("response+%i:", a))) %>%
-                pull()
-            
-            # Add prediction matrices and training end date
-            out$predict_x <- predict_x
-            out$predict_geo_values <- predict_geo_values
-            out$train_end_date <- train_end_date
-            
-            all_out[[paste0("ahead+", a)]] <- out
-        }
+        train_end_dates <- Reduce(c, lapply(ahead, 
+                                            function(a) get_train_end_date(lagged_df, a)))
```

Comment:
Also weird multiline behavior. Maybe a return before `lapply`? Can this be done with `sapply` instead? Or `purrr::map_dbl() %>% lubridate::ymd()`?

Reply:
`apply` and `sapply` don't work here because they coerce dates into numerics. The `Reduce` code was a solution I found here (https://stackoverflow.com/questions/46357983/r-apply-function-returns-numeric-value-on-date-variables). I've replaced it with `purrr::map_dbl() %>% lubridate::as_date()`, perhaps that is more readable.